NIFI-8462 Refactored PutSyslog and ListenSyslog using Netty

exceptionfactory opened the pull request:
#### Description of PR

NIFI-8462 Refactors the `PutSyslog` and `ListenSyslog` Processors to use Netty for socket communication.  This approach replaces the use of custom socket channel dispatcher classes in `nifi-processor-utils`.  Leveraging the Netty `SslHandler` also removes the use of custom `SSLSocket` and `SSLEngine` handling in `nifi-security-socket-ssl` classes.  This approach resolves latent issues with TLS 1.3 communication in these processors, described in NIFI-7468.

The new `nifi-event-transport` module abstracts communication handling for both UDP and TCP communication, providing interfaces and classes that can be leveraged in other components.  Refactoring other components to leverage Netty should eventually remove the need for error-prone `SSLEngine` handling in custom classes.

Changes also include refactored unit tests for `PutSyslog` and `ListenSyslog`, eliminating the use of mocked senders and removing unnecessary integration tests.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

ottobackwards commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/ByteArrayMessageNettyEventServerFactory.java:
```diff
@@ -0,0 +1,78 @@
+                    new DatagramByteArrayMessageDecoder(),
+                    byteArrayMessageChannelHandler
+            ));
+        } else {
```
would this be more re-usable if you could set the FrameDecoder?  Or would you be expected to re-implement the factory?

exceptionfactory replied:
That's a good question.  The initial purpose of this particular `EventServerFactory` implementation was to abstract the configuration details necessary for constructing an `EventServer` capable of receiving delimited byte arrays.  Netty provides a number of message decoders as well as the ability to create custom decoders capable of returning more structured objects.  I would expect different implementations of `NettyEventServerFactory` for different use cases.

ottobackwards commented:
If someone was going to use these new classes, it might be good to have a readme or something for this part of the code

ottobackwards commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSender.java:
```diff
@@ -0,0 +1,100 @@
+        try {
+            final Future<Channel> futureChannel = channelPool.acquire().sync();
+            final Channel channel = futureChannel.get();
+            try {
```
Can the channel not be active here?  Can it not be writeable?

exceptionfactory replied:
The `FixedChannelPool` configured in `NettyEventSenderFactory` includes a `ChannelHealthChecker` that checks whether the channel is active and connected.  That should minimize the possible types of exceptions thrown in this try block, but the finally block is intended to ensure that the `Channel` is always returned to the pool.

exceptionfactory commented:
> If someone was going to use these new classes, it might be good to have a readme or something for this part of the code

Thanks for the feedback @ottobackwards!  Providing some additional documentation would be helpful, although I think it would be better to wait until other components start leveraging this module.  I attempted to provide some high-level abstractions in the `EventSender` and `EventServer` interfaces, but even these interfaces and implementations are geared sending and receiving byte arrays for Syslog messages.  There should be some reuse of these classes for other processors, such as `PutTCP` and `ListenTCP`, among others.  If there are particular components that could use some additional description, feel free to point those out.  However, I expect some changes when it comes to leveraging similar approaches with other components.

Please let me know if you have any additional comments or questions!

gresockj commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutSyslog.java:
```diff
@@ -150,6 +149,7 @@
                     "messages will be sent over a secure connection.")
             .required(false)
             .identifiesControllerService(SSLContextService.class)
+            .dependsOn(PROTOCOL, TCP_VALUE)
```
I noticed that in ListenSyslog, the SSL Context Service property is always available, but in PutSyslog it's conditional on TCP being set.  Can you add the condition to both?

Also, if you have an SSL Context Service selected and then change it back to UDP, the processor is marked as invalid even though the SSL Context Service property is no longer available to be configured.  Is it possible to unset this property if protocol is set to UDP, or is that just a limitation of the current "dependsOn" functionality?

exceptionfactory replied:
Good catch, I had intended to update both processors to make SSL Context Service depend on TCP.

I noticed the issue with SSL Context Service and changing from TCP to UDP.  It seems like that is something that might make sense to address at the framework level.

gresockj commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ListenSyslog.java:
```diff
@@ -152,7 +145,7 @@
                 + "the <Message Delimiter> up to this configured maximum number of messages")
         .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
-        .expressionLanguageSupported(false)
+        .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
```
If you enable EL, you'll need to call evaluateAttributeExpressions when you use the property below.  Also, I'm not sure FLOWFILE_ATTRIBUTES is appropriate here, since ListenSyslog is annotated with INPUT_FORBIDDEN.

exceptionfactory replied:
Good catch, that should not have been changed, will revert back.

gresockj replied:
It might still make sense to let it be ExpressionLanguage.VARIABLE_REGISTRY if you want to allow for the additional flexibility, though it seems like a setting that wouldn't likely be externalized to the variable registry.

exceptionfactory replied:
Yes, that would be an option, but agree that there doesn't seem to be much value in making it configurable through the variable registry.

gresockj reviewed (changes requested):
This will be a nice upgrade, @exceptionfactory!  I tested it successfully with UDP and TCP with and without SSL Context Service, works as expected.  Just a few comments to address.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/message/ByteArrayMessage.java:
```diff
@@ -0,0 +1,39 @@
+ * Byte Array Message with Sender
+ */
+public class ByteArrayMessage {
+    private byte[] message;
```
Good point, will make the change.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/EventLoopGroupFactory.java:
```diff
@@ -0,0 +1,63 @@
+        this.workerThreads = workerThreads;
+    }
+
+    EventLoopGroup getEventLoopGroup() {
```
Good question, I originally intended for this base class and the inheriting subclasses to be scoped as narrowly as possible so that implementations would remain in this package.  Changing this to protected would maintain that intent while allowing a bit more flexibility, so will change to protected.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSenderFactory.java:
```diff
@@ -0,0 +1,163 @@
+
+    private Supplier<List<ChannelHandler>> handlerSupplier = () -> Collections.emptyList();
+
+    private final String address;
```
Sure, thanks for the suggestion.

exceptionfactory commented on nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSenderFactory.java:
```diff
@@ -0,0 +1,163 @@
+
+    private ChannelInitializer<Channel> getChannelInitializer() {
+        StandardChannelInitializer<Channel> channelInitializer;
+        if (sslContext == null) {
```
Sounds good, will update.

exceptionfactory commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutSyslog.java:
```diff
@@ -222,213 +224,97 @@ protected void init(final ProcessorInitializationContext context) {
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
         final int batchSize = context.getProperty(BATCH_SIZE).evaluateAttributeExpressions().asInteger();
-
         final List<FlowFile> flowFiles = session.get(batchSize);
```
That's a good point, I removed the yield when I removed the unnecessary pruning of idle senders, but it still makes sense to yield in the absence of Flow Files.

exceptionfactory commented:
Thanks for the thorough review and helpful feedback @gresockj!  I added a commit including the changes and rebased on main to incorporate recent updates.

thenatog commented:
Tested this out using the template found here: https://community.cloudera.com/t5/Community-Articles/NiFi-Send-to-syslog/ta-p/248638 and found that PutSyslog and ListenSyslog are working. Reviewed the code and it looks good.

thenatog commented:
+1 will merge.
